### ISSUE TITLE
Environment rollout should always be of shape (N, C, H, W) and Normalized

### ISSUE THREAD

**Issue by OleguerCanal** [CampusAI/Hamiltonian-Generative-Networks#26]
Now if I sample a random rollout from the environment it comes in shape (N, H, W, C) not Normalized.

I saw that you convert it into the right format into the dataloader but I don't think its the dataloader responability. For testing I may want to generate rollouts from the environment and directly feed them to the HGN.forward()

visualize_rollout should also take things in that format, so we can easily visualize network outputs as well.

**Comment by charlio23** [comment]
I will check it!
However, IMO I think that the dataloader should carry out the normalization. We have to think of an environment as a folder that contains images, so they should have their original format.

**Comment by OleguerCanal** [comment]
Actually, on a second thought i think we should not normalize data. Inputs being from 0 to 1 is ok for now, the outputs of the decoder are passed through a sigmoid so that should be ok right?

I think internally it can be implemented as a sequence of images but the output should be (B, N, C, H, W), so that it matches the input of the NN. B is batch, N is sequence length and C is channels.

Then the visualizer code which builds gifs should take np.arrays  of shape (N, C, H, W) where C can either be 1 or 3. This way we can also easily visualize HGN result outputs.

**Comment by charlio23** [comment]
The only inconvenience of reshaping the array into (N, C, H, W) is that the animator function takes arrays of shape (N, H, W, C) by default because matplotlib reads images as (H, W, C). [https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.imshow.html](url)

The way we should see the environment data generation is the same way as if we were reading images from a directory. When you read an image, you usually receive a numpy array of shape (H, W, C). Then your dataloader changes the array format into a torch tensor of the desired shape.

I understand that this is important in order to visualize the outputs from the HGN. Maybe we could add some code in the visualization function to convert from (C, H, W) to (H, W, C) automatically.

**Comment by OleguerCanal** [comment]
Good point.

What do you think of creating an utils.py script with functions to convert from one format to the other? Maybe we can also put the visualization functions there

**Comment by fedetask** [comment]
I agree with not normalizing the data, values should lie in [0, 1].

Okay maybe we just need two functions to go from (C, H, W) to (H, W, C) and vice versa

**Comment by OleguerCanal** [comment]
This way we can convert "raw" rollouts from the enviroment to network inputs without using the dataloader (can be useful to test certain initial conditions and such).

To visualize an HGN rollout now I'm doing something like this, would be nice to have a function which deals more automatically:
```python
        rollout = self.reconstructed_rollout.detach().numpy()
        rollout = np.squeeze(rollout, axis=0)
        rollout = np.array(np.split(rollout, len(self.q_s), axis=0))
        rollout = rollout.transpose((0, 2, 3, 1))
        rollout = np.array(250*rollout, dtype=np.uint8)  # Probably not needed step

        if (rollout.shape[-1] == 1):
            rollout = np.squeeze(rollout, axis=-1)
        visualize_rollout(rollout)
```

**Comment by stathius** [comment]
Not sure if you can normalize the data if they are created online.

**Comment by OleguerCanal** [comment]
Good point, we are not normalizing the inputs for now and things look good. Input are between 0 and 1.

**Comment by fedetask** [comment]
This issue was addressed in #30
